### PR TITLE
Native call tracing infrastructure + native build system overhaul

### PR DESCRIPTION
**Pull request by grendello**
Migrating code from https://github.com/xamarin/xamarin-android/pull/8800

This PR implements native method call tracing infrastructure based on
[libunwind](https://github.com/libunwind/libunwind/), initially used in
tracing the reasons for why `Blazor` apps break when marshal methods are
in use.  Utility of such infrastructure, however, is beyond just that
particular task, it should be available for general use by both us and
applications.

This PR does **not** enable the tracing support, it merely provides the
entire infrastructure to do so at a later date.  Part of that infrastructure
is the new layout of the native code build system:

  * All of the native bits we build are now placed under `src/native/`
  * A lot of code which was part of `src/monodroid/` is now in several
    directories under `src/native/`.
  * A single top-level `src/native/CMakeLists.txt` file together with
    CMake presets file (`CMakePresets.json.in`) is used to build all
    the target configurations
  * The presets file is processed by `xaprepare`, to create `CMakePresets.json`
    file which contains all the version-specific bits replacements such as
    Android API levels, paths to utilities etc.
  * One can create a `CMakeUserPresets.json` file in the same directory to
    provide one's own local modifications to various configs found in the 
    `CMakePresets.json` file. The user presets file must not be checked into
    git.

The subdivision of code previously found in `src/monodroid/` is due to desire
to have a single `src/native/` subdirectory per each **shared** library we
create as part of the build process (`libmonodroid.so`, `libxamarin-app.so`,
`libxamarin-debug-helper.so` and possibly others in the future).  The shared
libraries, however, share set of sources and those sources have been placed
in `src/native/` subdirectories, to produce **static** archives, subsequently
linked into the shared libraries.

This subdivision allows for nice separation of concerns, but also for easy
sharing of compilation flags, static source analysis (via `clang-check`), and
creates a flexible set of components which can be combined to produce the final
shared libraries with different sets of features/APIs/etc.

The hierarchy of `CMakeLists.txt` files is also easier to follow and modify than
it was previously.

### CODE REVIEW

File: build-tools/cmake/xa_macros.cmake [hunk 1]
```diff
@@ -200,7 +200,8 @@ function(xa_common_prepare)
     -fstack-protector-strong
     -fstrict-return
     -fno-strict-aliasing
-    -ffunction-sections
+    -fno-function-sections
```
Comment (jonpryor) [inline build-tools/cmake/xa_macros.cmake:203]: What does this *do*? Why add it?  From the [`clang -ffunction-sections` docs](https://clang.llvm.org/docs/ClangCommandLineReference.html#cmdoption-clang-fno-function-sections):

> Place each function in its own section

I thus infer that `-fno-function-sections` *doesn't* place each function into its own section?  I guess that could make the `.so` smaller?

Likewise the following line, [`clang -fno-data-sections`](https://clang.llvm.org/docs/ClangCommandLineReference.html#cmdoption-clang-fno-data-sections) implies that each data is *not* placed into its own section.

File: src/Xamarin.Android.Build.Tasks/Tests/Xamarin.ProjectTools/Resources/Base/BuildReleaseArm64SimpleDotNet.apkdesc [hunk 2]
```diff
@@ -98,5 +98,5 @@
       "Size": 1904
     }
   },
-  "PackageSize": 2689557
+  "PackageSize": 2742805
```
Comment (jonpryor) [inline src/Xamarin.Android.Build.Tasks/Tests/Xamarin.ProjectTools/Resources/Base/BuildReleaseArm64SimpleDotNet.apkdesc:101]: Increase by 53KB is somewhat odd, especially as I'm not sure what it's *from*.  Most changes appear negligible: `libmonosgen-2.0.so` is only ~5KB, `System.Private.CoreLib.dll` is only 4KB.

The change appears to be mostly due to `libmonodroid.so`, which is up by 122KB!

*That* may warrant further investigation.  *Why* is `libmonodroid.so` so much larger now?

File: src/native/libunwind/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,392 @@
...
+file(STRINGS "${CONFIGURE_AC}" UNWIND_EXTRA_VERSION_AC REGEX "^[ \t]*define[ \t]*\\([ \t]*pkg_extra,[ \t]*(.*)\\)")
+string(REGEX REPLACE "^[ \t]*define[ \t]*\\([ \t]*pkg_major,[ \t]*(.*)\\)" "\\1" UNWIND_MAJOR_VERSION "${UNWIND_MAJOR_VERSION_AC}")
+string(REGEX REPLACE "^[ \t]*define[ \t]*\\([ \t]*pkg_minor,[ \t]*(.*)\\)" "\\1" UNWIND_MINOR_VERSION "${UNWIND_MINOR_VERSION_AC}")
+string(REGEX REPLACE "^[ \t]*define[ \t]*\\([ \t]*pkg_extra,[ \t]*(.*)\\)" "\\1" UNWIND_EXTRA_VERSION "${UNWIND_EXTRA_VERSION_AC}")
+
+set(PKG_MAJOR "${UNWIND_MAJOR_VERSION}")
+set(PKG_MINOR "${UNWIND_MINOR_VERSION}")
+set(PKG_EXTRA "${UNWIND_EXTRA_VERSION}")
+set(PACKAGE_STRING "libunwind-xamarin")
+set(PACKAGE_BUGREPORT "")
```
Comment (jonpryor) [inline src/native/libunwind/CMakeLists.txt:20]: Why not set `PACKAGE_BUGREPORT`?

File: src/native/libunwind/fixes/aarch64/Gos-linux.c [hunk 4]
```diff
@@ -0,0 +1,145 @@
+/* libunwind - a platform-independent unwind library
```
Comment (jonpryor) [inline src/native/libunwind/fixes/aarch64/Gos-linux.c:1]: Given that this file is being *copied* into our repo, shouldn't it go into `src-ThirdParty`?

File: src/native/libunwind/fixes/aarch64/Gos-linux.c [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.  */
+
+//#include "unwind_i.h"
+#include "../../../../external/libunwind/src/aarch64/unwind_i.h"
```
Comment (jonpryor) [inline src/native/libunwind/fixes/aarch64/Gos-linux.c:28]: Isn't there a way to instead "alter" the include dirs so that the desired `unwind_i.h` is appropriately included?

That would certainly be cleaner than copying this file to make a one-line change.

File: src/native/monodroid/debug-constants.cc [hunk 6]
```diff
@@ -0,0 +1,10 @@
+#include "debug.hh"
+#include "shared-constants.hh"
+
+using namespace xamarin::android;
+using namespace xamarin::android::internal;
+
+extern "C" const char *__get_debug_mono_log_property (void)
```
Comment (jonpryor) [inline src/native/monodroid/debug-constants.cc:7]: Pity I didn't notice this in e99b02b788, but *generally* double-underscore `__` is reserved for the compiler + toolchain, and should not be used in "user" code (that's us! user code!).

I also don't understand why `DEBUG_MONO_LOG_PROPERTY` is "special" here in getting a `__get_` function + macro, vs. e.g. `DEBUG_MONO_ENV_PROPERTY` which is a `static inline constexpr`.

File: build-tools/cmake/xa_macros.cmake (hunk 1 above)
Comment (grendello) [reply build-tools/cmake/xa_macros.cmake:203]: In theory it can make the app/library smaller, but at a (slight) cost of speed of execution (functions might end up far from each other and force long instead of short calls, for instance) and it might impede LTO (link-time optimizations), so I opted for the "traditional" section sharing.

File: src/native/monodroid/pinvoke-override-api.cc [hunk 7]
```diff
@@ -363,6 +357,15 @@ _monodroid_lookup_replacement_method_info (const char *jniSourceType, const char
 	return JniRemapping::lookup_replacement_method_info (jniSourceType, jniMethodName, jniMethodSignature);
 }
 
+static void
+monodroid_log_traces (uint32_t kind, const char *first_line)
```
Comment (jonpryor) [inline src/native/monodroid/pinvoke-override-api.cc:361]: There's no mention of `monodroid_log_traces` in the commit message.  What's it _for_?

File: src/Xamarin.Android.Build.Tasks/Tests/Xamarin.ProjectTools/Resources/Base/BuildReleaseArm64SimpleDotNet.apkdesc (hunk 2 above)
Comment (grendello) [reply src/Xamarin.Android.Build.Tasks/Tests/Xamarin.ProjectTools/Resources/Base/BuildReleaseArm64SimpleDotNet.apkdesc:101]: I think it's the effect of two changes - the change to use static methods (which you note farther down) may cause the compiler to more eagerly inline code, and the removal of per-function sections may have caused LTO to inline more code.

File: src/native/libunwind/CMakeLists.txt (hunk 3 above)
Comment (grendello) [reply src/native/libunwind/CMakeLists.txt:20]: It's set to override the bug report address in `libunwind`, we don't want to have their address in our code base.

File: src/native/monodroid/xamarin_getifaddrs.cc [hunk 8]
```diff
@@ -876,7 +879,7 @@ get_link_address (const struct nlmsghdr *message, struct _monodroid_ifaddrs **if
 				}
 
 				if (payload_size > 0) {
-					size_t alloc_size = ADD_WITH_OVERFLOW_CHECK (size_t, payload_size, room_for_trailing_null);
```
Comment (jonpryor) [inline src/native/monodroid/xamarin_getifaddrs.cc:879]: the move away from the `ADD_WITH_OVERFLOW_CHECK()` macro to `Helpers::add_with_overflow_check<T>()` is also an unmentioned change.  Why is it part of this PR?

File: src/native/libunwind/fixes/aarch64/Gos-linux.c (hunk 4 above)
Comment (grendello) [reply src/native/libunwind/fixes/aarch64/Gos-linux.c:1]: No, because it's not part of the current release of the library - it's a local workaround for an issue, hopefully to be removed at some later date when the issue is fixed by upstream.

File: src/native/native.targets [hunk 9]
```diff
@@ -145,60 +143,53 @@
...
         Command="$(NinjaPath) -v"
-        WorkingDirectory="$(IntermediateOutputPath)%(AndroidSupportedTargetJitAbi.AndroidRID)-ubsan-Release"
-    />
+        WorkingDirectory="$(IntermediateOutputPath)%(AndroidSupportedTargetJitAbi.AndroidRID)-ubsan-Release" />
 
     <Touch Files="@(_BuildAndroidAnalyzerRuntimesOutputs)" />
   </Target>
 
   <Target Name="_CleanRuntimes"
-      AfterTargets="Clean">
```
Comment (jonpryor) [inline src/native/native.targets:193]: This original indentation was correct…

File: src/native/monodroid/debug-constants.cc (hunk 6 above)
Comment (grendello) [reply src/native/monodroid/debug-constants.cc:7]: Yeah, we shouldn't use `__` in macros, functions, classes etc, but it's "legacy" so I opted not to touch it.  As for the logic behind it? I have no idea, it's another "legacy" thing - I assume that this code was called by something else (either managed or native, likely the former) so I felt it safer to leave it alone.  And the macro calling this function is to have one location where the name is fetched, "just in case".

File: src/native/shared/CMakeLists.txt [hunk 10]
```diff
@@ -0,0 +1,77 @@
+set(LIB_NAME xa-shared-bits)
```
Comment (jonpryor) [inline src/native/shared/CMakeLists.txt:1]: Similar to the `runtime-base` question, what is `xa-shared-bits`?

Comment (jonpryor) [reply src/native/shared/CMakeLists.txt:1]: …and is there a good way to know when this is for a static lib vs. a shared lib vs.…

File: src/native/shared/log_functions.cc [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+#include <android/log.h>
+
+#include "java-interop-logger.h"
+#include "log_types.hh"
+
+// Must match the same ordering as LogCategories
+static constexpr std::array<const char*, 12> log_names = {
+	"*none*",
+	"monodroid",
+	"monodroid-assembly",
```
Comment (jonpryor) [inline src/native/shared/log_functions.cc:13]: The "duplication" between `LOG_CATEGORY_NAME_MONODROID_ASSEMBLY` and this feels odd, *especially* when `LOG_CATEGORY_NAME_MONODROID_ASSEMBLY` isn't used.

File: src/native/monodroid/pinvoke-override-api.cc (hunk 7 above)
Comment (grendello) [reply src/native/monodroid/pinvoke-override-api.cc:361]: There's no mention because it's not actually used here - this PR which sets up infrastructure for tracing.  The PR started this way (thus the branch name), but then I decided that it should only introduce "neutral" changes and prepare infrastructure for follow-up PRs which actually use the infrastructure (like Perfetto implementation)

It's part of code introduced in `src/native/monodroid/monodroid-tracing.cc`, also not really used by anything in this PR.

File: src/native/monodroid/xamarin_getifaddrs.cc (hunk 8 above)
Comment (grendello) [reply src/native/monodroid/xamarin_getifaddrs.cc:879]: It's not mentioned because it doesn't change anything in the way this code **works**, it's functionally equivalent to the macro.  The reason for his change is twofold - first, I much prefer functions than preprocessor macros, since they're typesafe.  Second, this way I can take advantage of the `source_location` header and report the code position in a much cleaner, type-safe and better way.  We had the macro before only because this was the only way to report on the `source:line` location (taking advantage of the `__FILE__` and `__LINE__` preprocessor macros, respectively)

File: src/native/runtime-base/CMakeLists.txt [hunk 12]
```diff
@@ -0,0 +1,61 @@
+set(LIB_NAME runtime-base)
```
[the remark this answers is not in the document]
Comment (grendello) [reply src/native/runtime-base/CMakeLists.txt:1]: Every subdirectory of `native` with `CMakeLists.txt` produces a `libXYZ.a` where `XYZ` is usually the subdirectory name.  `runtime-base` is used by three separate runtime libraries - `libmonodroid.so`, `libxamarin-native-tracing.so` and `libxamarin-debug-app-helper.so`

File: src/native/shared/CMakeLists.txt (hunk 10 above)
Comment (grendello) [reply src/native/shared/CMakeLists.txt:1]: Similar to `runtime-base`, it's a library that's shared by even more components than `runtime-base`

File: src/native/shared/log_functions.cc (hunk 11 above)
Comment (grendello) [reply src/native/shared/log_functions.cc:13]: Seeing as this PR was getting bigger and bigger, I didn't implement the changes I want to  make to all the `log*` functions in this file. Treat this as "infrastructure preparation" as well.

File: src/native/runtime-base/shared-constants.hh [hunk 13]
```diff
@@ -83,6 +98,19 @@ namespace xamarin::android::internal
 
 		// Documented in NDK's <android/log.h> comments
 		static constexpr size_t MAX_LOGCAT_MESSAGE_LENGTH = 1023;
+
+		static constexpr std::string_view LOG_CATEGORY_NAME_NONE               { "*none*" };
```
[the remark this answers is not in the document]
Comment (grendello) [reply src/native/runtime-base/shared-constants.hh:102]: They will replace the literals in `log_functions.cc` eventually

File: build-tools/cmake/xa_common.cmake [hunk 14]
```diff
@@ -1 +1,143 @@
...
+
+if(DEBUG_BUILD AND NOT DISABLE_DEBUG)
+  add_compile_definitions(DEBUG)
+endif()
+
+if(ANDROID_ABI MATCHES "^(arm64-v8a|x86_64)")
+  add_compile_definitions(ANDROID64)
+endif()
+
+if (ANDROID_NDK_MAJOR LESS 20)
```
Comment (jonpryor) [inline build-tools/cmake/xa_common.cmake:133]: Do we need to care about this anymore? We're on NDK *26* now!

File: src/native/libunwind/CMakeLists.txt (hunk 3 above)
Comment (jonpryor) [reply src/native/libunwind/CMakeLists.txt:20]: Let me rephrase: I'm not wondering why we're setting it at all.  I'm wondering why we're setting it to the empty string instead of github.com/xamarin/xamarin.android/issues.

Comment (grendello) [reply src/native/libunwind/CMakeLists.txt:20]: We're setting it to empty string because it sets bug report address for **libunwind**, not us - so we shouldn't put our address in their codebase (even though nobody's probably  going to see it there). Just a matter of keeping stuff tidy, I guess.

File: src/native/libunwind/fixes/aarch64/Gos-linux.c (hunk 4 above)
Comment (grendello) [reply src/native/libunwind/fixes/aarch64/Gos-linux.c:1]: https://github.com/libunwind/libunwind/issues/702

File: build-tools/cmake/xa_common.cmake (hunk 14 above)
Comment (grendello) [reply build-tools/cmake/xa_common.cmake]: The file is no longer used, I deleted it.

**Comment by jonpryor** [comment]
Draft commit message:

```markdown
ontext: https://github.com/xamarin/xamarin-android/pull/8800

Context: 8bc7a3e84f95e70fe12790ac31ecd97957771cb2
Context: 68368189d67c46ddbfed4e90e622f635c4aff11e
Context: https://github.com/libunwind/libunwind/issues/702

Begin adding native method call tracing infrastructure based on
[libunwind][0], initially used in tracing the reasons for why
[Blazor][1] apps break when LLVM marshal methods are enabled
(8bc7a3e8, 68368189).  The utility of such infrastructure, however,
is beyond just that particular task; it should be available for
general use by both us and applications.

Method call tracing support is not enabled by default.
To enable it, set the `$(_AndroidEnableNativeStackTracing)` MSBuild
property to `true`.  This will TODO-Grendel.

Additionally, change the directory layout for native code.

Previously, `src/monodroid` would produce *five* native libs:

  * `libmono-android.debug.so`
  * `libmono-android.release.so`
  * `libxa-internal-api.so`
  * `libxamarin-app.so`
  * `libxamarin-debug-app-helper.so`

This made for a large `src/monodroid/CMakeLists.txt`, complicating
maintenance.  We had considered trying to move these into separate
`src/LIBRARY-NAME` directories, but it makes things easier with
CMake if we introduce an intermediate directory.

Introduce a new `src/native` directory, to hold native binaries:

  * `src/native/monodroid`: new location for `libmono-android.*.so`

  * `src/native/xamarin-debug-app-helper`: source for
    `libxamarin-debug-app-helper.so`, containing code moved from
    previous `src/monodroid`

  * `src/native/xamarin-app-stub`: source for the "stub"
    `libxamarin-app.so`; the "real" one is produced in the app build.

Some `src/native/*` directories produce libraries (`.so` files). and
some produce *static* archives (`.a` files) to simplify using code
across native libraries.

A [cmake-presets file][2] is processed by `xaprepare`, to create a
`src/native/CMakePresets.json` which contains all the version-specific
bits replacements such as Android API levels, paths to utilities etc.

Developers can create a `src/native/CMakeUserPresets.json` file to
provide local modifications to config options within
`src/native/CMakePresets.json.in`.

Other changes:

  * `src/native/libunwind/fixes/aarch64/Gos-linux.c` is an altered
    copy of [`libunwind/src/aarch64/Gos-linux.c`][3] in order to
    workaround libunwind/libunwind#702.

  * *Begin* turning some instance member functions into static member
    functions, as we don't otherwise need instances of those classes.

  * Begin using `std::source_location` for better crash messages,
    instead of using `__FILE__` and `__LINE__`.

  * Fix endless recursion from a typo in `Util::ends_with()` overload.

[0]: https://github.com/libunwind/libunwind
[1]: https://learn.microsoft.com/aspnet/core/blazor/hybrid/tutorials/maui?view=aspnetcore-8.0
[2]: https://cmake.org/cmake/help/latest/manual/cmake-presets.7.html
[3]: https://github.com/libunwind/libunwind/blob/9cc4d98b22ae57bc1d8c253988feb85d4298a634/src/aarch64/Gos-linux.c
```
